python_tests/mpi_test.py

```diff
@@ -0,0 +1,61 @@
+# run it with
+# NVFUSER_DUMP=fusion_ir mpirun --allow-run-as-root -n 2 python mpi_nvfuser.py
+
+import torch
+import nvfuser
+from nvfuser.pytorch_utils import torch_dtype_to_nvfuser_dtype
+from torch import Tensor
+from typing import Tuple
+
+from nvfuser import FusionDefinition, DataType
+
+import os
+
+rank = int(os.environ["OMPI_COMM_WORLD_LOCAL_RANK"])
+
+torch.cuda.set_device(rank)
+
+### Inputs
+inputs = [
+    torch.randn(2, 4, device="cuda")[rank:rank+1, ...],
+]
+
+class MultiDeviceModel(FusionDefinition):
+
+    def definition(self):
+        # dynamic shape isn't working? at least I'm getting asserts
+        #self.t0 = self.from_pytorch(inputs[0])
+
+        self.t0 = self.define_tensor((2, 4), (False, False), dtype=DataType.Float)
+
+        # looks like I cannot have scalar inputs to any expression, hitting an assert
+        #self.s0 = self.define_constant(2.0)
+        #self.t1 = self.ops.mul(self.t0, self.s0)
+
+        # relu seems to also be complaining, I assume sharding operation has to be a set
```

Comment:
This should be supported though, let me look into it

But FYI, in the lowering a resharding "set" is supposed to be inserted before the relu in this case, si that the relu becomes non-resharding

Reply:
yeah, so sounds like this is something that should be done `automatically`?

I'm not saying that we "have to" do it this way, but this is one of the other thing that we'd want to establish at the API level. i.e. maybe we would be better served with a higher level API that encapsulate the mesh/parallel_type. (I think we have different view on what a high vs low level API is, but hopefully this one sounds somewhat clear :wink: )

Reply:
> yeah, so sounds like this is something that should be done `automatically`?

Yes, it supposed to be done automatically already, I need to check why it is not the case here

> I'm not saying that we "have to" do it this way, but this is one of the other thing that we'd want to establish at the API level. i.e. maybe we would be better served with a higher level API that encapsulate the mesh/parallel_type. (I think we have different view on what a high vs low level API is, but hopefully this one sounds somewhat clear 😉 )

Understood :) But I totally agree, and in this precise case I already implemented it to be automatic

Reply:
I just forgot to plumb into the MultiDeviceExecutor the function that automatically inserts the resharding sets. I will fix that